### ISSUE TITLE
pyright loses generic type association after isinstance

### ISSUE THREAD

**Issue by sherbang** [microsoft/pyright#6991]
**Describe the bug**

If a variable of a generic type is evaluated using `isinstance`, then pydantic no longer recognizes it being associated with the generic.

**Code or Screenshots**
```python
from typing import reveal_type


class Obj:
    ...


class Test[T]:
    def get_obj(self) -> T | Obj:
        ...

    def test(self) -> None:
        o = self.get_obj()
        reveal_type(o)  # Type of "o" is "T@Test | Obj"
        if isinstance(o, Obj):
            o = None
        reveal_type(o)  # Type of "o" is "object* | None"  (Expected: "T@Test | None")
```

**VS Code extension or command-line**

Both within VS Code, as well as the pyright command-line.
```
> pyright --version
pyright 1.1.346
```

**Comment by erictraut** [comment]
The current behavior is correct. The type that is reported as `object*` is, in effect, an intersection between `object` and `T@Test`. You can read about it [in this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables).
